src/Microsoft.Data.SqlClient/src/Microsoft/Data/SqlClient/SqlConnectionPoolKey.cs

```diff
@@ -79,49 +45,62 @@ internal string AccessToken
+
+        private SqlConnectionPoolKey(SqlConnectionPoolKey key) : base(key)
         {
-            get
-            {
-                return _clientCertificateRetrievalCallback;
-            }
+            _credential = key.Credential;
+            _accessToken = key.AccessToken;
+            _serverCertificateValidationCallback = key._serverCertificateValidationCallback;
+            _clientCertificateRetrievalCallback = key._clientCertificateRetrievalCallback;
+            CalculateHashCode();
         }
 
-        internal SqlClientOriginalNetworkAddressInfo OriginalNetworkAddressInfo
+        internal SqlConnectionPoolKey(string connectionString,
+                                    SqlCredential credential,
+                                    string accessToken,
+                                    ServerCertificateValidationCallback serverCertificateValidationCallback,
+                                    ClientCertificateRetrievalCallback clientCertificateRetrievalCallback,
+                                    SqlClientOriginalNetworkAddressInfo originalNetworkAddressInfo) : base(connectionString)
         {
-            get
-            {
-                return _originalNetworkAddressInfo;
-            }
+            Debug.Assert(_credential == null || _accessToken == null, "Credential and AccessToken can't have the value at the same time.");
+            _credential = credential;
+            _accessToken = accessToken;
+            _serverCertificateValidationCallback = serverCertificateValidationCallback;
+            _clientCertificateRetrievalCallback = clientCertificateRetrievalCallback;
+            _originalNetworkAddressInfo = originalNetworkAddressInfo;
+            CalculateHashCode();
         }
 
         public override bool Equals(object obj)
         {
-            SqlConnectionPoolKey key = obj as SqlConnectionPoolKey;
-
-            return (key != null &&
-                    _credential == key._credential &&
-                    ConnectionString == key.ConnectionString &&
-                    string.CompareOrdinal(_accessToken, key._accessToken) == 0 &&
-                    _serverCertificateValidationCallback == key._serverCertificateValidationCallback &&
-                    _clientCertificateRetrievalCallback == key._clientCertificateRetrievalCallback &&
-                    _originalNetworkAddressInfo == key._originalNetworkAddressInfo);
+            return (obj is SqlConnectionPoolKey key
+                && _credential == key._credential
+                && ConnectionString == key.ConnectionString
+                && string.CompareOrdinal(_accessToken, key._accessToken) == 0
+                && _serverCertificateValidationCallback == key._serverCertificateValidationCallback
+                && _clientCertificateRetrievalCallback == key._clientCertificateRetrievalCallback
+                && _originalNetworkAddressInfo == key._originalNetworkAddressInfo);
         }
```

Comment:
This function could be taken out to the common section; like [here](https://github.com/dotnet/SqlClient/pull/1292/commits/4d1f09530317d3ac5264b183ed1dd94db7af7041#diff-813dd48970098a3e3745790fdd8a33f0a9d98162936ebe899fbec804d91f19edR116).